File: assignment_0/install/lib/ur5_controller/sin_movement.py
#!/usr/bin/env python3
import rospy as ros
import numpy as np
from sensor_msgs.msg import JointState
from std_msgs.msg import Float64MultiArray


def talker():
    # Create a ros node 
    ros.init_node('mkthis_fcking_rbtmv', anonymous=True)
    # This is the initial position that you find in params.py
    q_des_q0 = np.array([-0.3223527113543909, -0.7805794638446351, -2.5675506591796875, -1.6347843609251917, -1.5715253988849085, -1.0017417112933558])
    # Specify the topic where to publish data
    pub_reduced_des_jstate = ros.Publisher("/ur5/joint_group_pos_controller/command", Float64MultiArray, queue_size=10)
    time = 0

    while not ros.is_shutdown():

        # Generate sin wave
        q_des = q_des_q0  + 0.1 * np.sin(2 * np.pi * 0.5 * time)
        print(q_des)

        # Send message
        msg = Float64MultiArray()
        msg.data = q_des
        pub_reduced_des_jstate.publish(msg)

        time = np.round(time + np.array([0.001]),  3)


if __name__ == '__main__':
    try:
        talker()
    except ros.ROSInterruptException:
        pass

File: assignment_0/src/ur5_controller/src/sin_movement.cpp
#include "ros/ros.h"
#include "std_msgs/Float64MultiArray.h"
#include <sensor_msgs/JointState.h>
#include <Eigen/Dense>


typedef Eigen::Matrix<double, 6, 1> JointStateVector;


int main(int argc, char** argv)
{
    // Initialize ROS Node
    ros::init(argc, argv, "sin_movement");
    ros::NodeHandle node;

    // Declare to publish to a topic and get a Publisher object (publisher destination joint state)
    ros::Publisher pub_des_jstate = node.advertise<std_msgs::Float64MultiArray>("/ur5/joint_group_pos_controller/command", 1000);

    // Variables
    double loop_frequency = 1000.;
    double loop_time = 0.;

    std_msgs::Float64MultiArray jointState_msg_array; 
    jointState_msg_array.data.resize(6);
    
    JointStateVector q_des = JointStateVector::Zero();
    JointStateVector q_des0 = JointStateVector::Zero();

    JointStateVector amp;
    JointStateVector freq;
    amp << 0.1, 0.1, 0.1, 0.1, 0.1, 0.1;
    freq << 0.2, 0.2, 0.2, 0.2, 0.2, 0.2;

    // How long will be the rate??
    ros::Rate loop_rate(loop_frequency);

    // Init value (find it in params.py)
    q_des0 << -0.3223527113543909, -0.7805794638446351, -2.5675506591796875, -1.6347843609251917, -1.5715253988849085, -1.0017417112933558;
    
    while (ros::ok())
    {
        // Generate sin wave
        q_des = q_des0.array() + amp.array() * (2 * M_PI * freq * loop_time).array().sin();

        // Create message object
        for (int i = 0; i < q_des.size(); i++)
        {
            jointState_msg_array.data[i] = q_des[i];
        }
        // Publish desired joint states message
        pub_des_jstate.publish(jointState_msg_array);
        
        loop_time += (double)1 / loop_frequency;
        ros::spinOnce();
        loop_rate.sleep();
    }

    return 0;
}